Client/SafeBrowsing/GSafeBrowsing.swift

```diff
@@ -0,0 +1,366 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import CommonCrypto
+import Shared
+
+private let log = Logger.browserLogger
+
+class SafeBrowsingClient {
+    private static let apiKey = "DUMMY_KEY"
+    private static let maxBandwidth = 2048 //Maximum amount of results we can process per threat-type
+    private static let maxDatabaseEntries = 250000 //Maximum amount of entries our database can hold per threat-type
+    private static let clientId = AppInfo.baseBundleIdentifier
+    private static let version = AppInfo.appVersion
+    
+    private let baseURL = "https://safebrowsing.brave.com"
+    private let session = URLSession(configuration: .ephemeral)
+    private let database = SafeBrowsingDatabase()
+    
+    public static let shared = SafeBrowsingClient()
+    
+    private init() {
+        self.database.scheduleUpdate { [weak self] in
+            self?.fetch({
+                if let error = $0 {
+                    log.error(error)
+                }
+            })
+        }
+    }
+    
+    func find(_ hashes: [String], _ completion: @escaping (_ isSafe: Bool, Error?) -> Void) {
+        database.find(hashes, completion: { [weak self] discoveredHashes in
+            guard let self = self else { return completion(true, nil) }
+            
+            if discoveredHashes.isEmpty {
+                return DispatchQueue.global(qos: .background).async {
+                    completion(true, nil)
+                }
+            }
+            
+            if !self.database.canFind() {
+                return DispatchQueue.global(qos: .background).async {
+                    completion(false, nil)
```

Comment:
I don't see the `database.canFind()` definition, but this seems to suggest that if there's a problem with the DB then we mark the URL as unsafe. Am I reading this right?

Reply:
I have resolved this and my PR should now 100% be complete with full features of safe-browsing. Cache has been fully implemented. Now we actually consult the cache right after a call to the DB to see if the URL is really safe. If it's not known, I no longer return a boolean. Instead, I return a state.

If the cache has previous history of the hash found in the db, it returns the cache's current state else it returns unknown if it cannot determine.. Maybe then we can warn the user that we don't know if the URL is good or bad?

Reply:
> Maybe then we can warn the user that we don't know if the URL is good or bad?

We're only allowed to warn users if we [know for sure that the URL is bad](https://developers.google.com/safe-browsing/v4/usage-limits#age-of-data-and-usage). If we're unsure, we can't show a warning.